org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/FoldFunction.java

```diff
@@ -0,0 +1,96 @@
+package org.metaborg.spoofax.shell.functions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+import org.metaborg.core.source.ISourceRegion;
+import org.metaborg.core.source.SourceRegion;
+import org.metaborg.spoofax.shell.output.FailOrSuccessResult;
+import org.metaborg.spoofax.shell.output.FoldResult;
+import org.metaborg.spoofax.shell.output.IResult;
+import org.metaborg.spoofax.shell.output.ISpoofaxTermResult;
+import org.spoofax.interpreter.terms.IStrategoTerm;
+
+/**
+ * Creates a {@link FoldResult} from a given {@link ISpoofaxTermResult}.
+ */
+public class FoldFunction implements FailableFunction<ISpoofaxTermResult<?>, FoldResult, IResult> {
+
+    /**
+     * Instantiate a {@link FoldFunction}.
+     * Explicit
+     */
+    public FoldFunction() {
+    }
+
+    @Override
+    public FailOrSuccessResult<FoldResult, IResult> apply(ISpoofaxTermResult<?> input) {
+
+        Helper helper = new Helper();
+
+        try {
+            IStrategoTerm term = input.ast().get();
+
+            term.writeAsString(helper, IStrategoTerm.INFINITE);
+
+            List<ISourceRegion> out = new ArrayList<>(helper.outputRegions);
+
+            return FailOrSuccessResult.successful(new FoldResult(term, out));
+        } catch (IOException e) {
+            // should not happen
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Helper to wrap {@link IStrategoTerm#writeAsString}.
+     */
+    private static class Helper implements Appendable {
```

Comment:
Documentation regarding why this class exists and what it does (it breaks lines on parentheses) is not clear enough. Possibly rename to better reflect what it does in the class name?